builder/actions/setup_ci_from_files.py

```diff
@@ -0,0 +1,301 @@
+                # It will assign environment_value to "SUCCESS".
+                #
+                # Valid JSON:
+                #    'input_role_arn': <AWS IAM role ARN here>
+                if ('input_role_arn' in item):
+                    try:
+                        arn_credentials = self.get_arn_role_credentials(str(item["input_role_arn"]))
+                        environment_value = "SUCCESS"
+                        environment_multi[environment_name + "_ACCESS_KEY"] = arn_credentials[0]
+                        environment_multi[environment_name + "_SECRET_ACCESS_KEY"] = arn_credentials[1]
+                        environment_multi[environment_name + "_SESSION_TOKEN"] = arn_credentials[2]
+                    except Exception as ex:
+                        sys.exit(
+                            f"[FAIL] {environment_name} [Input Role ARN]: Exception ocurred trying to get role ARN credentials")
+
+                # The same as "input_role_arn" but instead of taking the IAM role arn directly, it instead
+                # takes a AWS Secret Name and assumes the value in that secret is the IAM Role ARN to use and assume.
+                # Valid JSON:
+                #    'input_role_arn_secret': <AWS Secret Name containing IAM Role ARN here>
+                if ('input_role_arn_secret' in item):
+                    try:
+                        input_role_arn = self.env_instance.shell.get_secret(str(item['input_role_arn_secret']))
+                        arn_credentials = self.get_arn_role_credentials(input_role_arn)
+                        environment_value = "SUCCESS"
+                        environment_multi[environment_name + "_ACCESS_KEY"] = arn_credentials[0]
+                        environment_multi[environment_name + "_SECRET_ACCESS_KEY"] = arn_credentials[1]
+                        environment_multi[environment_name + "_SESSION_TOKEN"] = arn_credentials[2]
+                    except:
+                        sys.exit(
+                            f"[FAIL] {environment_name} [Input Role ARN Secret]: Exception ocurred trying to get role ARN credentials")
+
+                # Downloads the S3 file at the given URL and sets environment_value to the downloaded (temporary) file.
+                # Valid JSON:
+                #   'input_s3': <S3 URL Here>
+                if ('input_s3' in item):
+                    try:
+                        tmp_file = tempfile.NamedTemporaryFile()
+                        tmp_file.flush()
+                        self.tmp_file_storage.append(tmp_file)
+                        tmp_s3_filepath = tmp_file.name
+                        self.copy_s3_file(str(item['input_s3']), tmp_s3_filepath)
+                        environment_value = str(tmp_s3_filepath)
+                    except Exception as ex:
+                        sys.exit(f"[FAIL] {environment_name} [Input S3]: Exception ocurred trying to get S3 file")
```

Comment:
are users expected to have aws cli configured to be able to access account that owns the bucket? 
what if the user does not have those creds? 
who is putting files in s3 and how we are versioning them? I would make the argument that any configs outside of secrets should be in github

Reply:
Users are expected to have AWS CLI Configured with access to the bucket. If the account does not have access, then it will fail the builder action. This is the same for if the environment variable JSON/XML file is an S3 url: it assumes the user has access and will fail if there is an exception due to a lack of access.

For who is putting the files into S3: the idea is that our teams would do that. That way we can share the same configuration file across all 4 CRTs and 4 IoT SDKs when pointing to a common set of configuration options. For example, things like IoT certificates and private keys for unit-testing. Currently we share these files across the CRTs and SDKs, but how we get these files into the builder for unit testing vary completely by the CRT/SDK with no consistency. The idea is that we would have a single JSON/XML file that all the unit tests use across all 4 SDKs.

The reason for putting it in S3 instead of Secrets or GitHub is that way we can share it across repositories easier. We could put them in Secrets I suppose instead of S3 if that would be preferred.

______

Another reason we have `input_s3` is that some unit tests needs files that we cannot copy-paste into GitHub secrets or AWS Secrets due to their encoding. An example of this is PKCS12 keys and Java keystore files, as both have encoding that doesn't translate to straight text that can be inputted into AWS Secrets or GitHub secrets, so we have to download these files from S3.

Reply:
whats the proposed way people without access to our s3 are going to run builder? are we making those files public?

Reply:
a couple general thoughts. agree with the general idea for making tests more uniform, but i think we need to be really careful to avoid making it impossibly hard to run outside of ci
1. how are those resources in s3 created and maintained? relying on people manually uploading the files can be really flaky and not future proof (we've ran into similar issues with c-s3 before). I'm not against the idea in general, but i think any files uploaded to s3 need to be versioned and automatically generated as part of some release process (ex. be created on builder version release)
2.  can people run those tests without builder? outside of ci? builder should be a tool to make tests easier to run, but we should not make tests so complicated that its not feasible to run them without builder

Reply:
1. The resources are just dumped into CI, I didn't add versioning or anything like that. You do bring up a good point about versioning and otherwise making it part of the release pipeline, as otherwise it would be too easy to change something and accidentally break multiple CI process across multiple repos.

2. Yeah, the tests can be run outside of the builder so long as the environment variables are set. What we could do to help people running it locally is make a "local" JSON/XML file that has example variables that they can fill-in/replace to allow the tests to run locally. While I have not made such file yet (I'm still trying to get it to run successfully and fully in the Java CRT repo on a custom branch) this is something I can look at doing.

_________

I agree that we need to be careful with this and thank you for bringing up these points. I think it will help future proof the solution moving forward and it highlights a few areas that still need thought/development.